File: internal/shellwire/shellwire.go
// Package shellwire defines the end-to-end framing for the driver reverse shell
// (step 3 of the design in proposals/draft/driver-reverse-shell.md).
//
// Frames travel as binary WebSocket messages between the driver (which owns the
// PTY) and the operator's client. The server-side relay
// (internal/server/shellrelay) passes them through opaquely and never parses
// them — the framing is a contract between the two endpoints only. Both the
// driver leg (this repo, step 3) and the CLI client (step 5) import this codec
// so the wire format has a single definition.
//
// A frame is [1-byte type][payload]:
//
//	0x00 data   — raw PTY bytes (both directions; a PTY master is one stream)
//	0x01 resize — terminal size, two big-endian uint16s: rows then cols
//	0x02 exit   — shell exit code, one big-endian int32
//	0x03 ping   — keepalive, no payload
package shellwire

import (
	"encoding/binary"
	"fmt"
)

// Type is the one-byte frame discriminator.
type Type byte

const (
	// TypeData carries raw PTY bytes in either direction.
	TypeData Type = 0x00
	// TypeResize carries a new terminal size (rows, cols).
	TypeResize Type = 0x01
	// TypeExit carries the shell's exit code.
	TypeExit Type = 0x02
	// TypePing is a keepalive with no payload.
	TypePing Type = 0x03
)

// Frame is a decoded wire frame: a type and its raw payload.
type Frame struct {
	Type    Type
	Payload []byte
}

// Data encodes a data frame carrying raw PTY bytes.
func Data(b []byte) []byte {
	msg := make([]byte, 1+len(b))
	msg[0] = byte(TypeData)
	copy(msg[1:], b)
	return msg
}

// Resize encodes a resize frame from rows and cols.
func Resize(rows, cols uint16) []byte {
	msg := make([]byte, 5)
	msg[0] = byte(TypeResize)
	binary.BigEndian.PutUint16(msg[1:3], rows)
	binary.BigEndian.PutUint16(msg[3:5], cols)
	return msg
}

// Exit encodes an exit frame carrying the shell's exit code.
func Exit(code int) []byte {
	msg := make([]byte, 5)
	msg[0] = byte(TypeExit)
	binary.BigEndian.PutUint32(msg[1:5], uint32(int32(code)))
	return msg
}

// Ping encodes a keepalive frame.
func Ping() []byte {
	return []byte{byte(TypePing)}
}

// Parse decodes a wire message into a Frame. It errors on an empty message.
func Parse(msg []byte) (Frame, error) {
	if len(msg) == 0 {
		return Frame{}, fmt.Errorf("shellwire: empty frame")
	}
	return Frame{Type: Type(msg[0]), Payload: msg[1:]}, nil
}

// ResizeDims decodes the rows and cols from a resize frame's payload.
func (f Frame) ResizeDims() (rows, cols uint16, err error) {
	if f.Type != TypeResize {
		return 0, 0, fmt.Errorf("shellwire: not a resize frame: type %#x", byte(f.Type))
	}
	if len(f.Payload) != 4 {
		return 0, 0, fmt.Errorf("shellwire: resize payload must be 4 bytes, got %d", len(f.Payload))
	}
	rows = binary.BigEndian.Uint16(f.Payload[0:2])
	cols = binary.BigEndian.Uint16(f.Payload[2:4])
	return rows, cols, nil
}

// ExitCode decodes the exit code from an exit frame's payload.
func (f Frame) ExitCode() (int, error) {
	if f.Type != TypeExit {
		return 0, fmt.Errorf("shellwire: not an exit frame: type %#x", byte(f.Type))
	}
	if len(f.Payload) != 4 {
		return 0, fmt.Errorf("shellwire: exit payload must be 4 bytes, got %d", len(f.Payload))
	}
	return int(int32(binary.BigEndian.Uint32(f.Payload))), nil
}

File: webui/src/lib/shellwire.ts
// TypeScript port of internal/shell/shellwire/shellwire.go — the end-to-end
// framing for the driver reverse shell. Frames travel as binary WebSocket
// messages between the driver (which owns the PTY) and the operator's client.
// The server-side relay passes them through opaquely and never parses them, so
// this codec is a contract between the two endpoints only; it must stay byte-for-
// byte compatible with the Go definition.
//
// A frame is [1-byte type][payload]:
//
//   0x00 data   — raw PTY bytes (both directions; a PTY master is one stream)
//   0x01 resize — terminal size, two big-endian uint16s: rows then cols
//   0x02 exit   — shell exit code, one big-endian int32
//   0x03 ping   — keepalive, no payload

// Subprotocol is the WebSocket subprotocol version token negotiated on the
// attach leg (Sec-WebSocket-Protocol: xagent-shell.v1). It carries no
// credential — the browser authenticates via its cookie session.
export const SUBPROTOCOL = 'xagent-shell.v1'

// READ_LIMIT mirrors shellwire.ReadLimit (1 MiB). The browser has no equivalent
// of coder/websocket's SetReadLimit, but the constant is exported so callers can
// reason about the maximum frame the driver may send.
export const READ_LIMIT = 1 << 20

// FrameType is the one-byte frame discriminator, matching shellwire.Type.
export enum FrameType {
  Data = 0x00,
  Resize = 0x01,
  Exit = 0x02,
  Ping = 0x03,
}

// A decoded wire frame. Only the field relevant to `type` is populated: `data`
// for Data, `rows`/`cols` for Resize, `code` for Exit, nothing for Ping.
export interface Frame {
  type: number
  data?: Uint8Array
  rows?: number
  cols?: number
  code?: number
}

// encodeData encodes a data frame carrying raw PTY bytes: [0x00, ...bytes]. The
// result is backed by a plain ArrayBuffer (never a SharedArrayBuffer) so it
// satisfies WebSocket.send's BufferSource parameter.
export function encodeData(bytes: Uint8Array): Uint8Array<ArrayBuffer> {
  const msg = new Uint8Array(1 + bytes.length)
  msg[0] = FrameType.Data
  msg.set(bytes, 1)
  return msg
}

// encodeResize encodes a resize frame from rows and cols as two big-endian
// uint16s: [0x01, rows>>8, rows, cols>>8, cols].
export function encodeResize(rows: number, cols: number): Uint8Array<ArrayBuffer> {
  const msg = new Uint8Array(5)
  const view = new DataView(msg.buffer)
  msg[0] = FrameType.Resize
  view.setUint16(1, rows & 0xffff, false)
  view.setUint16(3, cols & 0xffff, false)
  return msg
}

// parse decodes a wire message into a Frame. It throws on an empty message,
// mirroring shellwire.Parse. Resize dimensions and exit codes are decoded eagerly
// so callers switch on `type` and read the matching field.
export function parse(msg: Uint8Array): Frame {
  if (msg.length === 0) {
    throw new Error('shellwire: empty frame')
  }
  const type = msg[0]
  const payload = msg.subarray(1)
  switch (type) {
    case FrameType.Resize: {
      if (payload.length !== 4) {
        throw new Error(`shellwire: resize payload must be 4 bytes, got ${payload.length}`)
      }
      const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
      return { type, rows: view.getUint16(0, false), cols: view.getUint16(2, false) }
    }
    case FrameType.Exit: {
      if (payload.length !== 4) {
        throw new Error(`shellwire: exit payload must be 4 bytes, got ${payload.length}`)
      }
      const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
      return { type, code: view.getInt32(0, false) }
    }
    case FrameType.Ping:
      return { type }
    default:
      return { type, data: payload }
  }
}
